### ISSUE TITLE
R template of a shared_ptr results in undefined class

### ISSUE THREAD

**Issue by AndLLA** [swig/swig#2372]
Setup:
- fresh installation of QuantLib 1.27 + QuantLib-SWIG 1.27 
- R 4.2.1
- swig 4.02 (as distributed with - Ubuntu 22.04)

Error:
```
Error in getClass(Class, where = topenv(parent.frame())) : 
  “_p_ext__shared_ptrT_InterpolatedDiscountCurveTLogLineart_t” is not a defined class
```

Sample R code for reproducing the bug:
```
library(QuantLib)

bToday = QuantLib::Date(31, "August", 2022)
QuantLib::Settings_setEvaluationDate(self = QuantLib::Settings_instance(), d = bToday)
print(Settings_instance()$getEvaluationDate())

bufDates = QuantLib::DateVector()
QuantLib::DateVector_append(self = bufDates, x = Date(31, "August", 2022))
QuantLib::DateVector_append(self = bufDates, x = Date(30, "September", 2022))
QuantLib::DateVector_size(self = bufDates)

bCurve = QuantLib::DiscountCurve(dates = bufDates , discounts = c(1.0, 1.0), dayCounter = Actual360())
```

The same error occurs in several other cases, all sharing a configuration similar to this one:
```
%{
using QuantLib::InterpolatedDiscountCurve;
%}

%shared_ptr(InterpolatedDiscountCurve<LogLinear>);

template <class Interpolator>
class InterpolatedDiscountCurve : public YieldTermStructure {
  public:
    InterpolatedDiscountCurve(const std::vector<Date>& dates,
                              const std::vector<DiscountFactor>& discounts,
                              const DayCounter& dayCounter,
                              const Calendar& calendar = Calendar(),
                              const Interpolator& i = Interpolator());
    const std::vector<Time>& times() const;
    const std::vector<Real>& data() const;
    const std::vector<Date>& dates() const;
    const std::vector<DiscountFactor>& discounts() const;
    #if !defined(SWIGR)
    std::vector<std::pair<Date,DiscountFactor> > nodes() const;
    #endif
};

%template(DiscountCurve) InterpolatedDiscountCurve<LogLinear>;

```

since everything works ok in python, 
i would assume it's not related to the specific quantlib configuration,
but rater to swig to the R translation layer ?

Thanks

**Comment by wsfulton** [comment]
Hi @AndLLA, this one for the QuantLib project to look at. If it is a SWIG bug, then please post a small standalone example here.

**Comment by AndLLA** [comment]
Hallo @wsfulton,
I hope I managed to replicate the symptom in the proper way.

Inside the attached zip file you can find:
 - MyLib.hpp: simple c++ header with a template class (OperationResult), along the lines of InterpolatedDiscountCurve above
 - MyLib.swig; the swig configuration, similar to the one reported above
 - setup.py: configuration used for building the python module

The python module is built as follows:
```
swig -c++ -python MyLib.swig
python3 setup.py build_ext --inplace
```

Afterwards the python module is successfully tested as follows:
```
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import MyLib
>>> MyLib.
MyLib.OpResAdd(          MyLib.OpResMult(         MyLib.OpResSub(          MyLib.SHARED_PTR_DISOWN  
>>> MyLib.OpResAdd(1,2)
<MyLib.OpResAdd; proxy of <Swig Object of type 'std::shared_ptr< OperationResult< OpAdd > > *' at 0x7f514cf14630> >
>>> buf = MyLib.OpResAdd(1,2)
>>> buf.getResult()
3.0
```

The R library is built as follows, using the exactly same configuration as for the python module:
```
swig -c++ -r MyLib.swig
c++ -c MyLib_wrap.cxx -I/usr/share/R/include -fPIC
R CMD SHLIB -o list.so MyLib_wrap.c MyLib.hpp
```

Testing the R library fails in the same way as the QuantLib example:
```
R version 4.2.1 (2022-06-23) -- "Funny-Looking Kid"

> dyn.load("list.so")
source("list.R")
> buf = OpResAdd(1,3)
Error in getClass(Class, where = topenv(parent.frame())) : 
  “_p_std__shared_ptrT_OperationResultTOpAddt_t” is not a defined class
> 

```

Thanks


[MyLib.zip](https://github.com/swig/swig/files/9628916/MyLib.zip)

**Comment by lballabio** [comment]
Confirmed, @wsfulton  — this is the shortest self-contained interface file I could write to reproduce the error:

```
%module MyLib

%include <std_shared_ptr.i>

%{
#include <memory>
template <class T> class TestClass {};
%}

%shared_ptr(TestClass<int>);

template <class T>
class TestClass {
  public:
    TestClass();
};

%template(TestClassI) TestClass<int>;
```

As @AndLLA, I get:
```
$ R -q
> library(MyLib)
> TestClassI()
Error in getClass(Class, where = topenv(parent.frame())) : 
  "_p_std__shared_ptrT_TestClassTintt_t" is not a defined class
>
```

The problem seems to be the interaction of `template` and `shared_ptr` — either one works file on its own.

**Comment by lballabio** [comment]
Further details: I wouldn't know where and what to change to fix it (I'm kind of hoping someone here does 😄 ), but the problem is that the constructor function for `TestClassI` in the generated R file says:
```
else new("_p_std__shared_ptrT_TestClassTintt_t", ref=ans);
```
while in the rest of the generated R file and in the generated C++ file, the name of the wrapped type is `_p_std__shared_ptrT_TestClassT_int_t_t`.  If I manually fix the type name in the call to `new` and rebuild, the module works.

**Comment by AndLLA** [comment]
hallo,
hopefully this can help,
the problem seems to be around `manglestr_default`.

As you can see from the printout below,
whenever `TestClass<(int)>` contains `(int)` is correctly translated to `_int_`,
while in the case of this bug, the function receives `TestClass<int>` which contains `<int>`
and therefore seems to be translated into the wrong mangled type.

```
************* scoerceout smartname <-- std::shared_ptr< TestClass<int> > 
manglestr_default 01 p. 
----------- getRClassName std::shared_ptr< TestClass<int> > 1 1 <-- p.std::shared_ptr< TestClass<int> > 
----------- getRClassName std::shared_ptr< TestClass<int> > 1 1 --> _p_std__shared_ptrT_TestClassTintt_t 
************* scoerceout smartname --> _p_std__shared_ptrT_TestClassTintt_t 
manglestr_default 01 p. 
----------- getRClassName p.TestClass<(int)> 0 0 <-- p.TestClass<(int)> 
----------- getRClassName p.TestClass<(int)> 0 0 --> _p_TestClassT_int_t 
```

**Comment by wsfulton** [comment]
@lballabio's example is fixed now, see #2386, so closing as fixed.
